agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/StartupDiagnostics.java

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.microsoft.applicationinsights.agent.internal.init;
+
+import com.microsoft.applicationinsights.agent.bootstrap.diagnostics.PidFinder;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.Optional;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+
+class StartupDiagnostics {
+
+  public static final String APPLICATIONINSIGHTS_DEBUG_RSS_ENABLED =
+      "applicationinsights.debug.rss.enabled";
+
+  // Execute with -XX:NativeMemoryTracking=summary
+  private static final String APPLICATIONINSIGHTS_DEBUG_NATIVE_MEM_TRACKING_ENABLED =
+      "applicationinsights.debug.native-mem-tracking.enabled";
+
+  public static final String APPLICATIONINSIGHTS_DEBUG_DIAG_EXPORT_TO_FILE =
+      "applicationinsights.debug.diag-export-to-file";
+  private final Logger startupLogger;
+
+  public StartupDiagnostics(Logger startupLogger) {
+    this.startupLogger = startupLogger;
+  }
+
+  static class DiagnosticsReport {
+
+    StringBuilder diagnosticBuilder = new StringBuilder();
+
+    void addDiagnostic(String diagnostic) {
+      diagnosticBuilder.append(diagnostic);
+      diagnosticBuilder.append(System.lineSeparator());
+      diagnosticBuilder.append(System.lineSeparator());
+    }
+
+    boolean isEmpty() {
+      return diagnosticBuilder.length() == 0;
+    }
+
+    @Override
+    public String toString() {
+      return diagnosticBuilder.toString();
+    }
+  }
+
+  void execute() {
+
+    DiagnosticsReport diagnosticsReport = new DiagnosticsReport();
+
+    if (Boolean.getBoolean(APPLICATIONINSIGHTS_DEBUG_RSS_ENABLED)) {
+      String os = System.getProperty("os.name");
+      if (os.equals("Linux")) {
+        String residentSetSize = findResidentSetSize();
+        diagnosticsReport.addDiagnostic(residentSetSize);
+      }
+    }
+
+    if (Boolean.getBoolean(APPLICATIONINSIGHTS_DEBUG_NATIVE_MEM_TRACKING_ENABLED)) {
+      String nativeSummary = executeNativeMemoryDiag();
+      diagnosticsReport.addDiagnostic(nativeSummary);
+    }
+
+    generateReport(diagnosticsReport);
+  }
+
+  private void generateReport(DiagnosticsReport diagnosticsReport) {
+    if (!diagnosticsReport.isEmpty()) {
+      startupLogger.info("Start-up diagnostics" + File.separator + diagnosticsReport);
+      boolean exportToFile = Boolean.getBoolean(APPLICATIONINSIGHTS_DEBUG_DIAG_EXPORT_TO_FILE);
+      if (exportToFile) {
+        saveIntoFile(diagnosticsReport);
+      }
+    }
+  }
+
+  private String findResidentSetSize() {
+    try (Stream<String> lines = Files.lines(Paths.get("/proc/self/status"))) {
+      Optional<String> optionalRss = lines.filter(line -> line.startsWith("VmRSS")).findAny();
+      return optionalRss.orElse("");
```

Comment:
I don't find the named var helps with readability here, but ok to keep if you do
```suggestion
      return lines.filter(line -> line.startsWith("VmRSS")).findAny().orElse("");
```